src/functions.php

```diff
@@ -33,93 +33,81 @@ function backendProvider(array $config): Backend
  * @param callable $callback
  * @return array
  */
-function download(Backend $backend, callable $callback=null): array
+function download(Backend $backend, $substitutes, callable $callback=null): array
 {
     $backend->setProgress($callback);
+    $backend->setSubstitutes($substitutes);
     return $backend->getVcards();
 }
 
 /**
- * Download images from CardDAV server
+ * uploading image files via ftp to the fritzbox fonpix directory
  *
- * @param array $cards
- * @return int
+ * @param $vcards     array     downloaded vCards
+ * @param $config     array
+ * @return                      number of transfered files
  */
-function downloadImages(Backend $backend, array $cards, callable $callback=null): array
+function uploadImages(array $vcards, $config)
 {
-    foreach ($cards as $card) {
-        if (isset($card->photo)) {
-            $uri = $card->photo;
-            $image = $backend->fetchImage($uri);
-            $card->photo_data = utf8_encode($image);
-
-            if (is_callable($callback)) {
-                $callback();
+    $user     = $config['user'];
+    $password = $config['password'];
+    $url      = $config['url'];
+    $path     = $config['fonpix'];
+    
+    $options = array('ftp' => array('overwrite' => true));
+    $context = stream_context_create($options);
+    $i = 0;
+    
+    foreach ($vcards as $vcard) {
+        if (isset($vcard->rawPhoto)) {                                     // skip all other vCards
+            if (preg_match("/JPEG/", strtoupper($vcard->photoData))) {     // Fritz!Box only accept jpg-files
+                $imgFile = imagecreatefromstring($vcard->rawPhoto);
```

Comment:
Ich versteh den Ablaluf nicht. Warum erst eine Imageressource erzeugen um die danach wieder in einen String zu verwandeln? Kann man das nciht komplett einsparen?

Reply:
Ich verstehe deine Frage nicht ganz.
Ich vermute einmal folgendes:
du fragst, warum nicht dass imagefile gleich beim download erzeugt wird - richtig. Könnte man machen - funktioniert dann für all die Server, die verlinkte Base64 Daten liefern. Aber nicht für Server, die von sich aus die Daten eingebettet liefern. Ich weiß nicht ob und wenn ja welche CardDAV-Server das tun - laut Spezifikation ist das für Verion 3.0 und 4.0 aber erlaubt und vorgesehen. Daher habe ich das embedding so gestrikt, als ob es von vornherein vom Server so gekommen wäre.
Beantwortet das deine Frage?

Reply:
macht nix, können wir separat nochmal anschauen